### ISSUE TITLE
Proposal for e2e configuration parameters.

### ISSUE THREAD

**Issue by timothysc** [kubernetes/kubernetes#18099]
**Problem Statement:**

Currently the e2e tests have too many hard coded and command line parameters for testing options. 

**Proposal:**

Centralize the configuration of the parameters into a single configuration object that can be easily: 

- serialized (json) 
- auto-generated with sane defaults
- user editable 
- hierarchical based on test
```
e2e: {
    gloablparam1 : "foo"
    testA: { 
          test_param1
    }
    density: { 
          max_pods_per_node
    }
}
```

/cc @kubernetes/sig-scalability @wojtek-t  @ixdy @jayunit100 @gmarek 
xref: #18097 , subsumes: #10479

**Comment by wojtek-t** [comment]
SGTM - I think we can use ConfigData objects once they are done.

**Comment by jayunit100** [comment]
Great!
In the meanwhile, I'm adding some configurability ... should we continue adding cmd switches? Or just use something more silent like env vars which won't effect as much code ...

**Comment by marun** [comment]
How will configuration be associated with consuming tests?  Given that ginkgo's use of descriptive text as test identifiers would likely be unnecessarily verbose, is the intent be to simply make up identifiers (as is currently done with cli parameters) that the appropriate test would know to consume?

**Comment by jayunit100** [comment]
The individual tests will define data structures hierarchichally that will be presented to the user in hierarchichal style.  So the tests, by definition, will know how to consume any and all parameters because they will be explicitly creating their own parameters.  

I guess We will need to possibly update the way flags work to express this hierarchichal configuration the way git/docker/and other similar hierharchichal config programs do.

**Comment by zmerlynn** [comment]
cc @kubernetes/goog-testing

**Comment by timothysc** [comment]
In discussion in the SIG today, we discussed whether this conflicts with ConfigData and the assertion was no.  To elaborate:  The only requirement for the e2e tests to run is a valid api-server endpoint.  It could be real or fake, so having parameters specified via config file would be an advantage in cases where not all moving parts are in place.   It doesn't force the requirement of ConfigData being present.

**Comment by ihmccreery** [comment]
This sounds great to me.  I'd like to see a more thorough design document before serious work begins, so that different teams can weigh in with their needs, etc.  This will be a pretty involved effort, especially if we intend to follow through, (and not leave with some command-line flags, some environment variables, and some data structures).

cc @spxtr specifically, as your work simplifying `e2e.sh` and `e2e.go` is closely coupled with this work.

**Comment by timothysc** [comment]
This stem'd as a desired thing to do given the plethora of magic numbers, but given other constraints, we never had it prioritized.  

/cc @spiffxp @countspongebob

**Comment by timothysc** [comment]
It seems like we can go-bindata the defaults then have an optional input argument as an over-ride.

**Comment by timothysc** [comment]
Jay is back now, and is going to start working on this soon.

**Comment by ixdy** [comment]
cc @kubernetes/sig-testing

**Comment by jayunit100** [comment]
Found the iniflags tool, seems ideal.

- iniflags: A simple config file.  we can support test hierarchies via `section` s. 
- compatible w/ existing golang `flag` library

**Comment by timothysc** [comment]
@spxtr @ixdy ^ any opinions?

**Comment by jayunit100** [comment]
A possible alternative if folks dont like `ini` is https://github.com/spf13/viper  @spiffxp .  opinions?  anyone have experience with it?

**Comment by spiffxp** [comment]
iniflags looks pragmatically fine, I was just asking from the perspective that since we sling around JSON and YAML for most other kubernetes resource specs, why not attempt to do the same here?  

spf13/viper made me go "ooo shiiiny" since we use spf13/cobra for CLI parsing as-is, and it's got optional integration with viper, but I personally have no hands-on experience with either

**Comment by ixdy** [comment]
I think @fejta has opinions. ref https://github.com/kubernetes/test-infra/pull/250 and https://github.com/kubernetes/test-infra/pull/257.

**Comment by timothysc** [comment]
viper looks gr8 imo.

**Comment by jayunit100** [comment]
viper working updated pr

**Comment by fejta** [comment]
So the idea here is we use ini files to set flags instead of bash? This seems like a strong idea to me.

**Comment by spxtr** [comment]
I would prefer json or yaml over ini, but I prefer all of the above over bash.

**Comment by jayunit100** [comment]
json + yaml + anything you want if we use `viper` :)

**Comment by timothysc** [comment]
https://github.com/kubernetes/kubernetes/issues/31453

**Comment by timothysc** [comment]
closes via #31030, #32902 

todo items would be to fill in for the other tests.
